docs/proposal/20211207-delegated-storage.md

```diff
@@ -0,0 +1,554 @@
+
+    // TypeInstance locking 
+    lockedBy(typeInstanceID, additionalParameters?) (string, error)
+    onLock(typeInstanceID, additionalParameters?) error
+    onUnlock(typeInstanceID, additionalParameters?) error
+    ```
+
+  An implementation of such service may vary between two use cases:
+
+  1. CRUD operations on output TypeInstance actually manages external resource (e.g. Vault) -> onCreate, onUpdate, and onDelete actually creates, updates and deletes a given resource.
+  1. output TypeInstance represents external resources managed in different way (e.g. via Capact actions - like Helm Runner). IMO we shouldn't move actual Helm release installation to TypeInstance "constructor").
+
+      - The service can also implement watch for external resources (e.g. Kubernetes secrets) and call `createTypeInstances` and `deleteTypeInstances` Hub mutations. We may provide Go framework to speed up such development, similarly as we have with Runner concept.
+
+1. The service could be implemented using one of the following solutions (but it is not limited to):
+
+  - [Dapr secrets](https://docs.dapr.io/developing-applications/building-blocks/secrets/secrets-overview/)
+  - [Kubernetes external secrets](https://github.com/external-secrets/kubernetes-external-secrets)
+  - [vault-k8s](https://github.com/hashicorp/vault-k8s)
+  - [db](https://upper.io/v4/getting-started/)
+
+## GraphQL API
+
+The new GraphQL API can be used both on CLI and UI.
+
+To list all available StorageBackends in Hub:
+
+```graphql
+query {
+   types(filter: { pathPattern: "cap.core.hub.storage.*" }) {
+        name
+        prefix
+        path
+    }
+}
+```
+
+To list all configured StorageBackends in Capact:
+
+```graphql
+# Ideally, but it could be too complicated:
+query ListTypeInstancesWithTypeRefFilter {
+  typeInstances(
+    filter: { typeRef: { path: "cap.core.hub.storage.*" } } # queries public Hub to fetch all Types attached to `cap.core.hub.storage` and return all TypeInstances which are of one of these TypeRefs
+  ) {
+    ...TypeInstance
+  }
+}
+
+# Alternatively: introduce `cap.core.attribute.hub.storage.backend` Attribute and simply do:
+
+query ListTypeInstancesWithAttributesAndTypeRefFilter {
+  typeInstances(
+    filter: {
+      attributes: [
+        { path: "cap.core.attribute.hub.storage.backend", rule: INCLUDE }
+      ]
+    }
+  ) {
+    ...TypeInstance
+  }
+}
+```
+
+To see the value for all TypeInstances, query:
+
+```graphql
+query ListTypeInstances {
+  typeInstances {
+    id
+    typeRef {
+      path
+      revision
+    }
+    lockedBy # resolver which calls proper backend storage service to ask for lock status
+    latestResourceVersion {
+      resourceVersion
+      createdBy
+      metadata {
+        attributes {
+          path
+          revision
+        }
+      }
+      spec {
+        value # resolver which calls proper backend storage service to ask for a given ResourceVersion value
+        additionalParameters
```

Comment:
why this directly under `spec`? it's a concept only for storage TI

Reply:
Agree 👍 Addressed